src/main/java/newbank/server/commands/Command.java

```diff
@@ -1,27 +1,94 @@
 package newbank.server.commands;
 
 import newbank.server.CustomerID;
-import newbank.server.exceptions.CommandInvalidSyntaxException;
-import newbank.server.exceptions.RequestNotAllowedException;
+import newbank.server.NewBank;
 
-/** Abstract representation of a command. */
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Abstract representation of a command.
+ */
 public abstract class Command {
-  public abstract String execute() throws CommandInvalidSyntaxException;
+  protected final NewBank bank;
+  protected final String[] tokens;
+  protected final CustomerID customerID;
+  protected List<Supplier<Optional<String>>> responsibilityChain;
+
+  protected Command(NewBank bank, String[] tokens, CustomerID customerID) {
+    this.bank = bank;
+    this.tokens = tokens;
+    this.customerID = customerID;
+  }
+
+  public abstract String execute();
 
   /**
    * @return the command syntax
    */
-  public String getSyntax() {
-    return "";
+  public abstract String getSyntax();
+
+  /**
+   * Apply each of the functions in the responsibility chain. If any of them
+   * return a String (containing a failure message), stop processing.
+   *
+   * @return `Optional`ly either a string containing a failure message, or null.
+   */
+  protected Optional<String> applyResponsibilityChain() {
+    return responsibilityChain.stream()
+        .map(Supplier::get)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .findFirst();
+  }
+
+  protected Optional<String> requestingHelp() {
```

Comment:
What is the value of optional Strings? I believe that for the String data structure we simply concatenate them to obtain the very same effect?

Reply:
Or, more specifically, in this case: we can just operate on an empty string?

Reply:
I'm very familiar with the concept of `Maybe`/`Optional` in other languages. However, here it seems to be equally dangerous (if not more?) as a standard type that implies the possibility of a `null`?

https://softwareengineering.stackexchange.com/a/309137

> Unfortunately, most languages that now offer Optional haven't abolished null, so you can only profit from the concept by instituting a strict policy of "absolutely no null, ever". Therefore, Optional in e.g. Java is not as compelling as it should ideally be.